presto-main/src/main/java/com/facebook/presto/cost/CoefficientBasedCostCalculator.java

```diff
@@ -92,77 +91,53 @@ public Visitor(Session session, Map<Symbol, Type> types)
         }
 
         @Override
-        protected Void visitPlan(PlanNode node, Map<PlanNode, PlanNodeCost> context)
+        protected PlanNodeCost visitPlan(PlanNode node, List<PlanNodeCost> context)
         {
-            visitChildren(node, context);
-            context.put(node, UNKNOWN_COST);
-            return null;
+            return UNKNOWN_COST;
         }
 
         @Override
-        public Void visitOutput(OutputNode node, Map<PlanNode, PlanNodeCost> context)
+        public PlanNodeCost visitOutput(OutputNode node, List<PlanNodeCost> context)
         {
-            visitChildren(node, context);
-            context.put(node, context.get(node.getSource()));
-            return null;
-        }
-
-        private void visitChildren(PlanNode node, Map<PlanNode, PlanNodeCost> context)
-        {
-            for (PlanNode source : node.getSources()) {
-                source.accept(this, context);
-            }
+            return Iterables.getOnlyElement(context);
         }
 
         @Override
-        public Void visitFilter(FilterNode node, Map<PlanNode, PlanNodeCost> context)
+        public PlanNodeCost visitFilter(FilterNode node, List<PlanNodeCost> context)
         {
-            boolean fullyEnforced = false;
-            if (node.getSource() instanceof TableScanNode) {
-                 fullyEnforced = visitTableScan((TableScanNode) node.getSource(), context, node.getPredicate());
-            }
-            else {
-                visitChildren(node, context);
-            }
-
-            PlanNodeCost sourceCost = context.get(node.getSource());
-            final double filterCoefficient = fullyEnforced ? 1.0 : FILTER_COEFFICIENT;
-            PlanNodeCost filterCost = sourceCost
-                    .mapOutputRowCount(value -> value * filterCoefficient);
-            context.put(node, filterCost);
-            return null;
+            PlanNodeCost sourceCost = Iterables.getOnlyElement(context);
+            return sourceCost
+                    .mapOutputRowCount(value -> value * FILTER_COEFFICIENT);
         }
 
         @Override
-        public Void visitProject(ProjectNode node, Map<PlanNode, PlanNodeCost> context)
+        public PlanNodeCost visitProject(ProjectNode node, List<PlanNodeCost> context)
         {
-            return copySourceCost(node, context);
+            return Iterables.getOnlyElement(context);
         }
 
         @Override
-        public Void visitJoin(JoinNode node, Map<PlanNode, PlanNodeCost> context)
+        public PlanNodeCost visitJoin(JoinNode node, List<PlanNodeCost> context)
         {
-            visitChildren(node, context);
-            PlanNodeCost leftCost = context.get(node.getLeft());
-            PlanNodeCost rightCost = context.get(node.getRight());
+            checkState(context.size() == 2);
```

Comment:
Preferably extract method
`checkSourceCostsCount(int expectedCount, List<PlanNodeCost> context)` and throw `checkArgument` with nice error message from it.